### DISCUSSION TITLE
Pattern match directly in function parameters

### DISCUSSION THREAD

**Discussion by jaschdoc** [flix/flix#8505] [Ideas]
We could allow pattern matches for singleton enums as function parameters, e.g.,

```flix
pub def fun(MyEnum.MyEnum(a,b,c): MyEnum): ... = // use a or b or c
```
Even better, we could completely drop the type annotation in the signature since we know the type directly from the syntax.
```flix
pub def fun(MyEnum.MyEnum(a,b,c)): ... = // use a or b or c
```
This would just be syntactic sugar for
```flix
pub def fun(t: MyEnum): ... = match t {
    case MyEnum.MyEnum(a, b, c) => // use a or b or c
}
```

This be a little more ergonomic than writing `pub def fun(t: ...): ... = let MyEnum.MyEnum(a, b, c) = t; ...` in beginning of functions.
Structs also offer some flexibility since the field access syntax is slightly less cumbersome than pattern matches, but it comes at the cost of functions having effects in the same region as the struct.

What do you think of this? I'm not sure if I like it or not :)

**Comment by JonathanStarup** [comment]
I like it - but the leftout type part is not easy and requires inference i think. It also introduces unnamed type parameters
```
enum E[t, ef](Unit -> t \ ef)
def map(E(f)): ...
// mustBePure(f)
```

**Comment by magnus-madsen** [comment]
I think this is difficult to implement.
